crud: add readview support (#372)
tarantool/crud: README.md

(see [tarantool/metrics#334](https://github.com/tarantool/metrics/issues/334)),
thus this feature will be unsupported for `metrics` driver.

### Read view

> A read view is an in-memory snapshot of data on instances that isn’t affected by future data modifications. Read views allow you to retrieve data using the `read_view_object:select()` and `read_view_object:pairs()` operations.

Reviewer: ```suggestion
A read view is an in-memory snapshot of data on instance that isn’t affected by future data modifications. Read views allow you to retrieve data using the `read_view_object:select()` and `read_view_object:pairs()` operations.
```

Read views can be used to make complex analytical queries. This reduces the load on the main database and improves RPS for a single Tarantool instance.

Read views have the following limitations:

  * Only the memtx engine is supported.

>   * Read view can be used starting from Tarantool Enterprise v2.11.0.

Reviewer: Let's add one more crud-specific point.

* There is no clusterwide readview support. For a sharded cluster, we open a readview on each storage. Due to a cluster's distributed nature, it is not guaranteed that they will open simultaneously.

#### Creating a read view

To create a read view, call the `crud.readview()` function.

```lua

> local foo = crud.readview(opts)

Reviewer: Let's use more meaningful name like `rv = crud.readview(opts)` everywhere

```

where:

* `opts`:
  * `name` (`?string`) - name of the read view
  * `timeout` (`?number`) - `vshard.call` timeout (in seconds)

**Example:**

```lua
local foo = crud.readview({name = 'foo', timeout = 3})
```

#### Closing a read view

When a read view is no longer needed, close it using the `read_view_object:close()` method because a read view may consume a substantial amount of memory.

```lua

> local foo = foo.readview()

Reviewer: ```suggestion
local foo = crud.readview()
```

foo:close(opts)
```

where:

* `opts`:
  * `timeout` (`?number`) - `vshard.call` timeout (in seconds)

> Otherwise, a read view is closed implicitly when the read view object is collected by the Lua garbage collector.

Reviewer: ```suggestion
A read view is also closed when the object is collected by the Lua garbage collector.
```

**Example:**

```lua
local foo = crud.readview()
foo:close({timeout = 3})
```

#### Read view select

`read_view_object:select()` supports multi-conditional selects, treating a cluster as a single space, same as `crud.select`.

```lua
local foo = crud.readview()
local objects, err = foo:select(space_name, conditions, opts)
foo:close()
```

where:

* `space_name` (`string`) - name of the space
* `conditions` (`?table`) - array of [select conditions](#select-conditions)

> * `opts`:

Reviewer: Let's better make a link to select opts and list which options are not supported here.

  - [2, 401, 'Mary', 46]
  - [3, 2804, 'David', 33]
  - [4, 1161, 'William', 81]
  - [5, 1172, 'Jack', 35]
  - [6, 1064, 'William', 25]
  - [7, 693, 'Elizabeth', 18]
- null
...
foo.close()
```

##### Read view select conditions

Select conditions for `read_view_object:select()` are the same as [select conditions](#select-conditions) for `crud.select`.

**Example:**

```lua
foo = crud.readview()
foo:select('customers', {{'<=', 'age', 35}}, {first = 10})
---
- metadata:
  - {'name': 'id', 'type': 'unsigned'}
  - {'name': 'bucket_id', 'type': 'unsigned'}
  - {'name': 'name', 'type': 'string'}
  - {'name': 'age', 'type': 'number'}
  rows:
  - [5, 1172, 'Jack', 35]
  - [3, 2804, 'David', 33]
  - [6, 1064, 'William', 25]
  - [7, 693, 'Elizabeth', 18]
  - [1, 477, 'Elizabeth', 12]
...
foo.close()
```

#### Read view pairs

You can iterate across a distributed space using the `read_view_object:pairs()` method.
Its arguments are the same as [`crud.readview.select`](#read-view-select) arguments except
`fullscan` (it does not exist because `crud.pairs` does not generate a critical
log entry on potentially long requests) and negative `first` values aren't
allowed.

> User could pass use_tomap flag (false by default) to iterate over flat tuples or objects.

Reviewer: ```suggestion
User could pass `use_tomap` flag (`false` by default) to iterate over flat tuples or objects.
```
